mona_openai/analysis/refusal.py

```diff
@@ -0,0 +1,112 @@
+# In order to compute sentence embeddings, we use SentenceTransformers framework
+# (https://www.sbert.net/).
+# More information about the different models (and the tradeoff between speed and
+# performance of each one can be found here:
+# https://www.sbert.net/docs/pretrained_models.html)
+EMBEDDING_MODEL_NAME = environ.get("EMBEDDING_MODEL_NAME", "all-MiniLM-L6-v2")
+
+# Example sentences of LLM refusal answers (we will compare a given text to each one of
+# them in order to compute the probability that the given text is a refusal answer).
+REFUSAL_EXAMPLES = [
+    "I'm sorry, but I cannot assist with providing or discussing .",
+    "I'm unable to provide information or guidance on .",
+    "I cannot provide assistance or information on illegal activities, including but "
+    "not limited to ."
+    "I am prohibited from generating explicit, adult, or inappropriate content",
+    "I cannot help you with generating or facilitating plagiarism or any form of "
+    "cheating",
+    "I cannot provide personalized legal advice or guidance on complex legal issues",
+    "I'm not able to assist with"
+    "I cannot provide assistance or guidance on .",
+    "As an AI language model, I am unable to fulfill requests that involve sharing "
+    "personal identifiable information, generating explicit content, providing "
+    "assistance for hacking or malicious activities, engaging in unlawful actions, or "
+    "promoting harm, discrimination, or illegal content.",
+]
+
+embeddings_model = None
+refusal_embeddings = None
+
+_DECIMAL_PLACES = 2
+
+
+def _lazy_model_init() -> None:
+    """
+    Complete
+    """
+    global embeddings_model, refusal_embeddings
+
+    # The embedding model is already initialize.
+    if embeddings_model is not None and refusal_embeddings is not None:
+        return
+
+    # This is the first call to this function - need model initialization.
+    try:
+        embeddings_model = SentenceTransformer(EMBEDDING_MODEL_NAME)
+    except RepositoryNotFoundError:
+        raise InvalidEmbeddingsModelName(
+            f"Given embeddings model name '{EMBEDDING_MODEL_NAME}' is invalid)"
+        )
+    # Get embeddings of refusal answers examples.
+    refusal_embeddings = [
+        _get_text_embedding(refusal_example)
+        for refusal_example in REFUSAL_EXAMPLES
+    ]
+
+
+def _get_text_embedding(text: str) -> Tensor:
+    """
+    Return sentence embedding of the given text.
+    """
+    return embeddings_model.encode(text, convert_to_tensor=True)
+
+
+def _get_embeddings_similarity(
+    embedding1: Tensor, embedding2: Tensor
+) -> float:
+    """
+    Return the cosine similarity between the two given sentence embeddings.
+    """
+    return util.pytorch_cos_sim(embedding1, embedding2).item()
+
+
+def _predict_refusal_prob(text: str) -> float:
+    """
+    Given a text, return the probably that it's a refusal answer by a LLM model.
+    """
+    # Compute the embedding for the given text.
```

Comment:
This and the below comments seem pretty redundant to me. The code is very clear here.